Automattic/woocommerce-payments
# Implementation of Intent request classes

Fixes #5057

#### Changes proposed in this Pull Request
The following requests have been added using request class:

- Create_Intent
- Update_Intent
- Get_Intent

The examples can be found below.

To start with this request, create a new Create Intent request and fill the data.

Example:

**Create intent**

```php
$currency                 = strtolower( $order->get_currency() );
$$customer_id             = $request->get_param( 'customer_id' );
$metadata                 = $request->get_param( 'metadata' ) ?? [];
$metadata['order_number'] = $order->get_order_number();

$wcpay_server_request = Create_Intent::create();
$wcpay_server_request->set_currency_code( $currency );
$wcpay_server_request->set_amount( WC_Payments_Utils::prepare_amount( $order->get_total(), $currency ) );
if ( $customer_id ) {
	$wcpay_server_request->set_customer( $customer_id );
}
$wcpay_server_request->set_metadata( $metadata );
$wcpay_server_request->set_payment_method_types( $this->get_terminal_intent_payment_method( $request ) );
$wcpay_server_request->set_capture_method( 'manual' === $this->get_terminal_intent_capture_method( $request ) );
$intent = $wcpay_server_request->send( 'create_wcpay_terminal_intent_request', $order );
```
The `create_wcpay_terminal_intent_request` filter have been added where someone could use it according to their needs.

**Update intent**

```php
$request = Update_Intention::create();
$request->set_intent_id( $payment_intent_id );
$request->set_currency_code( strtolower( $currency ) );
$request->set_amount( WC_Payments_Utils::prepare_amount( $amount, $currency ) );
$request->set_metadata( $this->get_metadata_from_order( $order, $payment_type ) );
$request->set_level3( $this->get_level3_data_from_order( $order ) );
$request->set_selected_upe_payment_method_type( (string) $selected_upe_payment_type, $this->get_payment_method_ids_enabled_at_checkout( null, true ) );
if ( $payment_country ) {
	$request->set_payment_country( $payment_country );
}
if ( true === $save_payment_method ) {
	$request->setup_future_usage();
}
if ( $customer_id ) {
	$request->set_customer( $customer_id );
}
$updated_payment_intent = $request->send( 'wcpay_update_intention_request', $order, $payment_intent_id );
```

The `wcpay_update_intention_request` filter have been added where someone could use it according to their needs. In this example, order and `intent_id` have been passed.

**Get intent**

```php
$request = Get_Intent::create();
$request->set_intent_id( $intent_id );
$intent = $request->send( 'wcpay_get_intent_request', $order );
```

The `wcpay_get_intent_request` filter have been added where someone could use it according to their needs. In this example, order have been passed.

For get and update intents requests, the intent id can only be set once to prevent possible issues with data tempering.








#### Testing instructions

Make sure that terminal create intent endpoint works correctly.
-->

*

-------------------


**Post merge**
